site-src/geps/gep-718.md

```diff
@@ -0,0 +1,278 @@
+	//
+	// If only one backend is specified and it has a weight greater than 0, 100%
+	// of the traffic is forwarded to that backend. If weight is set to 0, no
+	// traffic should be forwarded for this entry. If unspecified, weight
+	// defaults to 1.
+	//
+	// Support: Extended
+	//
+	// +optional
+	// +kubebuilder:default=1
+	// +kubebuilder:validation:Minimum=0
+	// +kubebuilder:validation:Maximum=1000000
+	Weight *int32 `json:"weight,omitempty"`
+}
+```
+
+A summarized diff for the changes being proposed:
+
+```patch
+diff --git a/apis/v1alpha2/shared_types.go b/apis/v1alpha2/shared_types.go
+index 1215300..228f861 100644
+--- a/apis/v1alpha2/shared_types.go
++++ b/apis/v1alpha2/shared_types.go
+@@ -96,10 +96,7 @@ type GatewayReference struct {
+ 
+ // RouteForwardTo defines how a Route should forward a request.
+ type RouteForwardTo struct {
+-	// ServiceName refers to the name of the Service to forward matched requests
+-	// to. When specified, this takes the place of BackendRef. If both
+-	// BackendRef and ServiceName are specified, ServiceName will be given
+-	// precedence.
++	// BackendRef is a reference to a backend to forward matched requests to.
+ 	//
+ 	// If the referent cannot be found, the rule is not included in the route.
+ 	// The controller should raise the "ResolvedRefs" condition on the Gateway
+@@ -115,30 +112,16 @@ type RouteForwardTo struct {
+ 	// in the BackendPolicy resource. For custom backends, it is encouraged to
+ 	// add a semantically-equivalent field in the Custom Resource Definition.
+ 	//
+-	// Support: Core
+-	//
+-	// +optional
+-	// +kubebuilder:validation:MaxLength=253
+-	ServiceName *string `json:"serviceName,omitempty"`
+-
+-	// BackendRef is a reference to a backend to forward matched requests to. If
+-	// both BackendRef and ServiceName are specified, ServiceName will be given
+-	// precedence.
+-	//
+-	// If the referent cannot be found, the rule is not included in the route.
+-	// The controller should raise the "ResolvedRefs" condition on the Gateway
+-	// with the "DegradedRoutes" reason. The gateway status for this route should
+-	// be updated with a condition that describes the error more specifically.
+-	//
+-	// Support: Custom
++	// Support: Core (Kubernetes Service)
++	// Support: Custom (any other resource)
+ 	//
+ 	// +optional
++	// +kubebuilder:default={kind: "service"}
+ 	BackendRef *LocalObjectReference `json:"backendRef,omitempty"`
+ 
+ 	// Port specifies the destination port number to use for the
+-	// backend referenced by the ServiceName or BackendRef field.
+-	// If unspecified, the destination port in the request is used
+-	// when forwarding to a backendRef or serviceName.
++	// backend referenced by BackendRef field. Port is required when the backend
++	// is a Kubernetes Service.
+ 	//
+ 	// Support: Core
+ 	//
+```
+For Kubernetes Services, the API will look like the following:
+
+```yaml
+...
+forwardTo:
+- backendRef:
```

Comment:
Could we remove the `backendRef` field and just have those fields exist in the `RouteForwardTo` struct? 

You'd end up with this: 
```
forwardTo:
  - name: foo-service-v1
    port: 80
    weight: 80
  - name: foo-service-canary
    port: 80
    weight: 20 
---
forwardTo:
-  name: foo-v1
    kind: server
    group: networking.acme.io
    port: 80
    weight: 80
-  name: foo-v1-canary
    kind: server
    group: networking.acme.io
    port: 80
    weight: 20
```

Reply:
This is mentioned here (https://github.com/kubernetes-sigs/gateway-api/pull/719/files#diff-200562f21cd2e117d439d1a59d8b3c844217b91f3d6224f34222687e9a8fa5faR267) to move the port/weight fields into the `backendRef` struct. 

I'm suggesting something similar but just the opposite, moving everything into the `forwardTo`.

Reply:
@robscott @youngnick Do you know if this is a violation of the upstream API convention or not?
I lean towards an xRef containing only the reference (as API evolves, having such a standard makes the UX much better, envoy is a good example here) but happy to tweak this if that's helpful.

Reply:
This is a good question. I can't find anything in the k8s API conventions about this. @lavalamp would it ever be acceptable to include concepts like port or weight in an object reference? We're trying to figure out if we can flatten our API a bit. Essentially translating this:

```
forwardTo:
- backendRef:
    group: ""
    kind: Service
    name: foo
  port: 80
  weight: 10
```

to something like this:

```
backendRefs:
- group: ""
  kind: Service
  name: foo
  port: 80
  weight: 10
```

Reply:
I think additional fields should be fine. As long as they don't collide with any standard fields, or fields we're likely to add. I doubt object references will ever have "port" or "weight" fields so you should be good. @deads2k will probably agree.